### DISCUSSION TITLE
Question about coupling external codes in MOOSE

### DISCUSSION THREAD

**Discussion by Ethan-xj** [idaholab/moose#28287] [Q&A General] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hello everyone, I'm doing some coupling works based on MOOSE. In detail, I'm coupling two external codes and they are solved by their own solvers. If I want to use MOOSE to control the coupling and solving of external programs, which features of MOOSE should I utilize? Could you provide some recommendations? Thank you very much.

Ethan

**Comment by GiudGiud** [comment]
Hello

One of the ways to do this is to use the ExternalProblem class. There are not very many examples outside of moose. 

I think Cardinal used to use an external problem

**Reply by Ethan-xj** [reply]
OK Giud. But Cardinal doesn't have many documents about the code structure. So it's a bit hard for me to understand the coupling strategy of Cardinal.

**Reply by GiudGiud** [reply, marked as answer]
The main idea is that it is creating a new Problem class to order the solve steps
https://github.com/neams-th-coe/cardinal/blob/devel/src/base/CardinalProblem.C

Then there is custom data mappings between Nek/OpenMC data and MOOSE, which is passed between codes using user objects
https://github.com/neams-th-coe/cardinal/blob/devel/src/userobjects/NekBinnedSideIntegral.C

An alternate example is the Petsc external problem. This is a lot more hard-coded for a specific PETSc solve (which is arguably not the best practice)
There too a new Problem class orders the solve
https://github.com/idaholab/moose/blob/next/modules/external_petsc_solver/src/problems/ExternalPETScProblem.C

**Reply by Ethan-xj** [reply]
Thanks and I will take it as refer!

**Reply by Ethan-xj** [reply]
@GiudGiud Hi Giud. I have some other questions. Can MOOSE run an external program through a static library, or must it be done through a dynamic library? Currently, I have two external programs that need to run, but they have different compilation environments, and I'm not sure how to resolve this issue.

Ethan

**Reply by GiudGiud** [reply]
Both external problems need to couple in memory? You can't call them in a script?

I think both are options. Dynamic is definitely easier if they are moose apps

**Reply by Ethan-xj** [reply]
Yes in memory. I want to use MOOSE to control the transient solution for two programs.
And they are not moose apps. So, I am wondering if static libraries are less dependent on the environment at runtime, which might make them easier to use.

**Reply by GiudGiud** [reply]
Static libraries will grow the executable. I'm not sure if there are much differences with regards to depending on the environment. Do you mean you plan to compile on one machine and run on another?

I think standard practice in moose is to build and link dynamic libraries. You will find more examples for dynamic libraries
